source/slang/slang-compiler.cpp

```diff
@@ -149,28 +150,64 @@ namespace Slang
         {
             outputBinary.addRange(result.outputBinary.getBuffer(), result.outputBinary.getCount());
         }
+        else if (appendTo == ResultFormat::SharedLibrary)
+        {
+            SLANG_ASSERT(!"Trying to append to a shared library");
+        }
+    }
+
+    SlangResult CompileResult::getSharedLibrary(ComPtr<ISlangSharedLibrary>& outSharedLibrary)
+    {
+        if (format == ResultFormat::SharedLibrary && sharedLibrary)
+        {
+            outSharedLibrary = sharedLibrary;
+            return SLANG_OK;
+        }
+        return SLANG_FAIL;
     }
 
-    ComPtr<ISlangBlob> CompileResult::getBlob()
+    SlangResult CompileResult::getBlob(ComPtr<ISlangBlob>& outBlob)
     {
         if(!blob)
         {
             switch(format)
             {
-            case ResultFormat::None:
-            default:
-                break;
+                case ResultFormat::None:
+                default:
+                    break;
 
-            case ResultFormat::Text:
-                blob = StringUtil::createStringBlob(outputString);
-                break;
+                case ResultFormat::Text:
+                    blob = StringUtil::createStringBlob(outputString);
+                    break;
 
-            case ResultFormat::Binary:
-                blob = createRawBlob(outputBinary.getBuffer(), outputBinary.getCount());
-                break;
+                case ResultFormat::Binary:
+                    blob = createRawBlob(outputBinary.getBuffer(), outputBinary.getCount());
+                    break;
+                case ResultFormat::SharedLibrary:
```

Comment:
I don't feel like having this `SharedLibrary` case on `CompileResult` really makes sense.

The `CompileResult` type is supposed to be a thin wrapper around "the stuff we'd write to the file specified with `-o`". The `SharedLibrary` case here is meeting that contract, but it certainly isn't a thin wrapper anymore when it retains an on-disk file and also lazily allocates and fills in an in-memory blob.

This case seems to only exist to handle host-callable code, and it seems to me that the "host callable" case shouldn't be producing `CompileResult`s in the first place, since there is no meaningful "what to write for `-o`" in this case anyway.

For me it kind of calls into question the idea of treating "host callable" as a `SlangTarget`, since it has so little in common with other examples of targets. All the other targets are things that produce raw text/binary output (only) and can meaningfully be used from the command line with output paths specified using `-o`.

The "host callable" target doesn't produce any kind of (meaningful) "raw" output. It is the equivalent of `LoadLibrary` or `dlopen`, and it isn't a meaningful thing to say "I'd like to write the output of `dlopen` to disk..."

I'd like to find a design where we sidestep this `CompileResult` issue by not having the host-callable path produce `CompileResult`s at all. I *almost* think that the host-callable path should be disjoint from the existing targets (so it is a separate API), but I'll concede that there is an issue there where there are codegen option that we'd want to specify for the host-callable path and we'd want those to not duplicate the existing way to set those options for ordinary targets.

I guess I don't have a concrete proposal here, but I think there is a "code smell" thing here that shows we maybe don't have the user-facing API design 100% right.

Reply:
Ok I understand your concern and can look into.